### PR TITLE
Move erts IO Polling to dedicated threads

### PR DESCRIPTION
**Pull request by garazdawi**
This PR implements a solution to the problem described in https://bugs.erlang.org/browse/ERL-140.

The implementation moves all IO polling out from the main scheduler loops into separate dedicated thread(s). By default one thread is started to be the poller and all file descriptors (or waitable objects) are handled by it using a kernel-poll mechanism.
It is possible to tune the number of polling threads using the `+IOt` and `+IOPt` options. e.g. `+IOt 2` will start 2 polling threads, `+IOPt 50` will start 50% as many polling threads as there are schedulers.
It is possible to tune the number of poll-sets using the `+IOp` and `+IOPp` options. On Linux and BSD it is possible to have 1..t poll-sets configured, on all other OSs the number of poll-sets have to be equal to the number of poll-threads.

On operating systems that can do concurrent updates of the kernel poll-set (linux and bsd), the new implementation has no global locks that need to be taken when updating the poll-set. On operating systems that cannot do concurrent updates, there is still a lock per poll-set that has to be taken.

The `+K` flag no longer has any effect and is ignored. If the user wants to not use kernel-poll, this can now be achieved through a configure flag when building the emulator.

### Linux and BSD

The basic flow of operation for handling a driver_select call is as follows:
* driver_select is called in a linked-in driver in a scheduler thread
  * the fd is used to hash into a global array to find it's slot
    * the hash is protected by striped mutexes
  * the port that was interested in the fd event in the slot
  * the pollset is modified to reflect the new interest using `epoll_ctl(kp_fd, EPOLL_CTL_ADD, fd, { .events = EPOLLIN|EPOLLONESHOT, .data = fd })`
* During all this, the poll-threads are using `epoll_wait`
  * When the event comes one poll-thread will be woken (because we use ONESHOT)
  * It looks for the correct slot and locks it
  * Delivers the correct signal to the port which calls ready_input/output

### Solaris/Windows/non-kernel poll

Basically the same flow happens, the major difference is that the actual inserting into the poll-set is done by the poll-thread itself instead of by the scheduler threads.

### Benchmarking

I've done quite a bit of benchmarking, mainly to make sure that we don't get any regressions when using this approach. The throughput and latency figures look promising, through I don't really have access to any great benchmarks or hardware to run them on. So if you have any socket IO intensive benchmarks, please run them and report back with the good or bad results.

### Outstanding issues

There are a few testcases that still fail on various platforms due to testcase timing issues, and some new testcases have to be written for the new erts options.

Enjoy

### CODE REVIEW

**Comment by vans163** [comment]
Is it possible to lock the polling threads to certain logical cores?

**Comment by garazdawi** [comment]
@vans163, at the moment it is not possible to bind the polling threads. I've been thinking that it may be a good thing to make this possible, but it is not implemented yet.

File: erts/emulator/beam/erl_driver.h [hunk 1]
```diff
@@ -289,10 +281,7 @@ typedef struct erl_drv_entry {
 			 unsigned int *flags); /* Works mostly like 'control',
 						  a synchronous
 						  call into the driver. */
-    void (*event)(ErlDrvData drv_data, ErlDrvEvent event,
-		  ErlDrvEventData event_data);
-                                /* Called when an event selected by 
-				   driver_event() has occurred */
+    void *unused_event_callback;
```
Comment (mikpe) [inline erts/emulator/beam/erl_driver.h]: Nit: if your aim is to maintain binary compatibility then make this entry "void (*unused_event_callback)(void);", i.e. keep it a function pointer.  C does not guarantee that function pointers and void* have the same size.
